includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
+  curl_setopt($cURL, CURLOPT_HEADER, 1);
+  curl_setopt($cURL, CURLINFO_HEADER_OUT, true);
+  switch (strtoupper($type)) {
+    case 'POST':
+    curl_setopt($cURL, CURLOPT_POST, true);
+    break;
+    case 'PUT':
+    curl_setopt($cURL, CURLOPT_CUSTOMREQUEST, 'PUT');
+    break;
+  }
+  if ($content) {
+    curl_setopt($cURL, CURLOPT_POSTFIELDS, $content);
+  }
+  curl_setopt($cURL, CURLOPT_HTTPHEADER, $headers);
+  $response = curl_exec($cURL);
+  $header_size = curl_getinfo($cURL, CURLINFO_HEADER_SIZE);
+  $response_code = curl_getinfo($cURL, CURLINFO_RESPONSE_CODE);
+  $headerSent = curl_getinfo($cURL, CURLINFO_HEADER_OUT );
+  curl_close($cURL);
+  if ($response === false) {
+    $messages = array(
+      'message' => t("Unable to receive response from curl_exec()"),
+      'message_sub' => array(),
+      'type' => 'watchdog',
+      'severity' => WATCHDOG_ERROR,
+    );
+    watchdog($messages);
+  } else {
+    $headers = substr($response, 0, $header_size);
+    $body = substr($response, $header_size);
+    if (empty($expect) || (isset($expect[$response_code]) && empty($expect[$response_code]))) {
+      return $body;
+    } else if (isset($expect[$response_code])) {
+      $expectation = $expect[$response_code];
+      $json = json_decode($body, true);
+      if (is_array($json)) {
+        if (isset($json[$expectation])) {
+          return $json;
+        } else {
+          $messages = array(
+            'message' => t("JSON key from @endpoint (@expectation) missing . @response"),
+            'message_sub' => array(
+              '@endpoint' => $endpoint, 
+              '@expectation' => $expectation,
+              '@response' => $response,
+            ),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+          watchdog($messages);
+        }
+      } else {
+        $messages = array(
+          'message' => t("Expected JSON key from @endpoint but could not decode. @response"),
+          'message_sub' => array(
+            '@endpoint' => $endpoint,
+            '@response' => $response,
+          ),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+        watchdog($messages);
+      }
+    } else {
+      $messages = array(
+        'message' => t("Expected one of @expected; got a @response_code"),
+        'message_sub' => array(
+          '@expected' => implode(',', array_keys($expect)),
+          '@response_code' => $response_code,
+        ), 
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+      watchdog($messages);
+    }
+  }
+  return;
+}
+
+/**
+ * Uploads file to Panopto using AWS S3 Uploader
+ * @param string $upload_target The URL of upload target
+ * @param string $file_path The URL which is the target of the request
```

Comment:
Because this function may throw an `Exception` this should be documented in the PHPDoc.  It can be as simple as:
```
@throws Exception The S3Client may throw an exception which is unhandled by this method.
```